searchlib/src/vespa/searchlib/attribute/postingstore.cpp

```diff
@@ -647,86 +651,92 @@ PostingStore<DataT>::update_stat()
+        uint32_t clusterSize = getClusterSize(typeId);
+        if (clusterSize == 0) {
+            if (isBitVector(typeId)) {
+                BitVectorEntry *bve = getWBitVectorEntry(iRef);
+                RefType iRef2(bve->_tree);
+                if (iRef2.valid()) {
+                    assert(isBTree(iRef2));
+                    if (_store.getCompacting(iRef2)) {
+                        BTreeType *tree = getWTreeEntry(iRef2);
+                        auto ref_and_ptr = allocBTreeCopy(*tree);
+                        tree->prepare_hold();
+                        // Note: Needs review when porting to other platforms
+                        // Assumes that other CPUs observes stores from this CPU in order
+                        std::atomic_thread_fence(std::memory_order_release);
+                        bve->_tree = ref_and_ptr.ref;
+                    }
                 }
+                if (_store.getCompacting(iRef)) {
+                    auto new_ref = allocBitVectorCopy(*bve).ref;
+                    _bvs.erase(iRef.ref());
+                    _bvs.insert(new_ref.ref());
+                    ref = new_ref;
+                }
+            } else {
+                assert(isBTree(typeId));
+                assert(_store.getCompacting(iRef));
+                BTreeType *tree = getWTreeEntry(iRef);
+                auto ref_and_ptr = allocBTreeCopy(*tree);
+                tree->prepare_hold();
+                ref = ref_and_ptr.ref;
             }
-            if (!_store.getCompacting(ref)) {
-                return ref;
-            }
-            auto new_ref = allocBitVectorCopy(*bve).ref;
-            _bvs.erase(ref.ref());
-            _bvs.insert(new_ref.ref());
-            return new_ref;
         } else {
-            if (!_store.getCompacting(ref)) {
-                return ref;
-            }
-            BTreeType *tree = getWTreeEntry(iRef);
-            auto ref_and_ptr = allocBTreeCopy(*tree);
-            tree->prepare_hold();
-            return ref_and_ptr.ref;
+            assert(_store.getCompacting(iRef));
+            const KeyDataType *shortArray = getKeyDataEntry(iRef, clusterSize);
+            ref = allocKeyDataCopy(shortArray, clusterSize).ref;
         }
     }
-    if (!_store.getCompacting(ref)) {
-        return ref;
-    }
-    const KeyDataType *shortArray = getKeyDataEntry(iRef, clusterSize);
-    return allocKeyDataCopy(shortArray, clusterSize).ref;
 }
 
 template <typename DataT>
 void
 PostingStore<DataT>::compact_worst_btree_nodes()
 {
     auto to_hold = this->start_compact_worst_btree_nodes();
-    _dictionary.normalize_posting_lists([this](EntryRef posting_idx) -> EntryRef
-                                        {
-                                            move_btree_nodes(posting_idx);
-                                            return posting_idx;
-                                        });
+    std::vector<bool> filter(RefType::numBuffers());
```

Comment:
Please create a helper function to create the filter.

Reply:
Fixed.